akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+   * the given [[WriteConsistency consistency level]] and
+   * [[WriteConsistency#timeout timeout]].
+   *
+   * The `Update` was still performed locally and possibly replicated to some nodes.
+   * It will eventually be disseminated to other replicas, unless the local replica
+   * crashes before it has been able to communicate with other replicas.
+   */
+  final case class UpdateTimeout(key: String, request: Option[Any]) extends UpdateFailure
+  /**
+   * The [[Update]] could not fetch current value according to the given
+   * [[ReadConsistency readConsistency level]] and [[ReadConsistency#timeout timeout]].
+   */
+  final case class ReadFailure(key: String, request: Option[Any]) extends UpdateFailure
+  final case class InvalidUsage(key: String, errorMessage: String, request: Option[Any])
+    extends RuntimeException(errorMessage) with NoStackTrace with UpdateFailure {
+    override def toString: String = s"InvalidUsage [$key]: $errorMessage"
+  }
+  /**
+   * If the `modify` function of the [[Update]] throws an exception the reply message
+   * will be this `ModifyFailure` message. The original exception is included as `cause`.
+   */
+  final case class ModifyFailure(key: String, errorMessage: String, cause: Throwable, request: Option[Any])
+    extends RuntimeException(errorMessage, cause) with NoStackTrace with UpdateFailure {
+    override def toString: String = s"ModifyFailure [$key]: $errorMessage"
+  }
+
+  /**
+   * Send this message to the local `Replicator` to delete a data value for the
+   * given `key`. The `Replicator` will reply with one of the [[DeleteResponse]] messages.
+   */
+  final case class Delete(key: String, consistency: WriteConsistency) extends Command
+
+  sealed trait DeleteResponse {
+    def key: String
+  }
+  final case class DeleteSuccess(key: String) extends DeleteResponse
+  final case class ReplicationDeleteFailure(key: String) extends DeleteResponse
+  final case class DataDeleted(key: String)
+    extends RuntimeException with NoStackTrace with DeleteResponse {
+    override def toString: String = s"DataDeleted [$key]"
+  }
+
+  /**
+   * Get current number of replicas, including the local replica.
+   * Will reply to sender with [[ReplicaCount]].
+   */
+  final case object GetReplicaCount
+
+  /**
+   * Java API: The [[GetReplicaCount]] instance
+   */
+  def replicaInstance = GetReplicaCount
+
+  /**
+   * @see [[GetReplicaCount]]
+   */
+  final case class ReplicaCount(n: Int)
+
+  /**
+   * Notify subscribers of changes now, otherwise they will be notified periodically
+   * with the configured `notify-subscribers-interval`.
+   */
+  case object FlushChanges
+
+  /**
+   * Java API: The [[FlushChanges]] instance
+   */
+  def flushChangesInstance = FlushChanges
+
+  /**
+   * Marker trait for remote messages serialized by
+   * [[akka.cluster.ddata.protobuf.ReplicatorMessageSerializer]].
+   */
+  trait ReplicatorMessage extends Serializable
```

Comment:
Of course. My thinking went more along the lines of whether alternative serializers like avro or kryo somehow use this UID as well, and if someone wanted to provide their own serialization of these types then that could help them. In the end it does not matter whether we add this annotation or not, people who want to misunderstand the presence of `Serializable` will do so either way. But I don’t have a strong opinion on this particular point.